Quality: PowerShell Summit GH-3324 (#3339)
MicrosoftDocs/windows-powershell-docs: docset/winserver2022-ps/activedirectory/Get-ADComputer.md

# Get-ADComputer

## SYNOPSIS

Gets one or more Active Directory computers.

## SYNTAX

### Filter (Default)

```powershell
Get-ADComputer [-AuthType <ADAuthType>] [-Credential <PSCredential>] 
  -Filter  <String> [-Properties <String[]>] [-ResultPageSize <Int32>]
  [-ResultSetSize <Int32>] [-SearchBase <String>] [-SearchScope <ADSearchScope>]
  [-Server <String>] [<CommonParameters>]
```

### Identity

```powershell
Get-ADComputer [-AuthType <ADAuthType>] [-Credential <PSCredential>] 
  [-Identity] <ADComputer> [-Partition <String>] [-Properties <String[]>]
  [-Server <String>] [<CommonParameters>]
```

### LdapFilter

```powershell
Get-ADComputer [-AuthType <ADAuthType>] [-Credential <PSCredential>] -LDAPFilter <String>
  [-Properties <String[]>] [-ResultPageSize <Int32>] [-ResultSetSize <Int32>]
  [-SearchBase <String>] [-SearchScope <ADSearchScope>] [-Server <String>] 
  [<CommonParameters>]
```

## DESCRIPTION

> The **Get-ADComputer** cmdlet gets a computer or performs a search to retrieve multiple computers.

Reviewer: We always want to wrap cmdlet names with backticks so they render as code. Our [style guide][01] has additional formatting guidance you might find useful.

```suggestion
The `Get-ADComputer` cmdlet gets a computer or performs a search to retrieve multiple computers.
```

[01]: https://learn.microsoft.com/powershell/scripting/community/contributing/powershell-style-guide

> The _Identity_ parameter specifies the Active Directory computer to retrieve.

Reviewer: We always bold our parameter names when they're not prefixed with a `-`, like `-Identity`. When they're prefixed, we wrap them in backticks to render as code.

```suggestion
The **Identity** parameter specifies the Active Directory computer to retrieve.
```

You can identify a computer by its distinguished name, GUID, security identifier
(SID) or Security Accounts Manager (SAM) account name. You can also set the
parameter to a computer object variable, such as `$<localComputerobject>` or

> pass a computer object through the pipeline to the _Identity_ parameter.

Reviewer: ```suggestion
pass a computer object through the pipeline to the **Identity** parameter.
```

> To search for and retrieve more than one computer, use the _Filter_ or
> _LDAPFilter_ parameters. The _Filter_ parameter uses the PowerShell Expression
> Language to write query strings for Active Directory. PowerShell Expression
> Language syntax provides rich type conversion support for value types received
> by the _Filter_ parameter. For more information about the _Filter_ parameter
> syntax, type `Get-Help`
> [about_ActiveDirectory_Filter](/previous-versions/windows/server/hh531527(v=ws.10)).
> If you have existing Lightweight Directory Access Protocol (LDAP) query strings,
> you can use the _LDAPFilter_ parameter.

Reviewer: ```suggestion
To search for and retrieve more than one computer, use the **Filter** or **LDAPFilter** parameters.
The **Filter** parameter uses the PowerShell Expression Language to write query strings for Active
Directory. PowerShell Expression Language syntax provides rich type conversion support for value
types received by the **Filter** parameter. For more information about the **Filter** parameter
syntax, type `Get-Help`
[about_ActiveDirectory_Filter](/previous-versions/windows/server/hh531527(v=ws.10)). If you have
existing Lightweight Directory Access Protocol (LDAP) query strings, you can use the **LDAPFilter**
parameter.
```

> This cmdlet retrieves a default set of computer object properties. To retrieve
> additional properties use the _Properties_ parameter. For more information about
> the how to determine the properties for computer objects, see the _Properties_
> parameter description.

Reviewer: ```suggestion
This cmdlet retrieves a default set of computer object properties. To retrieve additional
properties use the **Properties** parameter. For more information about the how to determine the
properties for computer objects, see the **Properties** parameter description.
```

This command gets a specific computer showing all the properties.

### Example 2: Get all computers with a name starting with a particular string

```powershell
Get-ADComputer -Filter 'Name -like "User01*"' -Properties IPv4Address | 

>     Format-Table Name,DNSHostName,IPv4Address -AutoSize

Reviewer: Where possible, we want to have spaces between list items so they're easier to read.

```suggestion
    Format-Table Name, DNSHostName, IPv4Address -AutoSize
```

```

```Output
name        dnshostname            ipv4address
----        -----------            -----------
User01-SRV1 User01-SRV1.User01.com 10.194.99.181
User01-SRV2 User01-SRV2.User01.com 10.194.100.3
```

This command gets all the computers with a name starting with a particular

> string and shows the name, dns hostname, and IPv4 address.

Reviewer: ```suggestion
string and shows the name, DNS hostname, and IPv4 address.
```

### Example 3: Gets all computers that have changed their password in specific time frame

```powershell
$Date = [DateTime]::Today.AddDays(-90) 
Get-ADComputer -Filter 'PasswordLastSet -ge $Date' -Properties PasswordLastSet | 

>     Format-Table Name,PasswordLastSet

Reviewer: ```suggestion
    Format-Table Name, PasswordLastSet
```

This command gets all the computers that have changed their password in the last 90 days.

### Example 4: Get computer accounts in a specific location using an LDAPFilter

```powershell
Get-ADComputer -LDAPFilter "(name=*laptop*)" -SearchBase "CN=Computers,DC= User01,DC=com"
```

```Output
name
----
pattiful-laptop
davidche-laptop
```

This command gets the computer accounts in the location

> CN=Computers,DC=User01,DC=com that are listed as laptops by using an
> _LDAPFilter_.

Reviewer: Here we want to wrap the value in backticks because it's an exact value used in the prompt.

```suggestion
`CN=Computers,DC=User01,DC=com` that are listed as laptops by using an
**LDAPFilter**.
```

### Example 5: Get all computer accounts using a filter

```powershell
Get-ADComputer -Filter *
```

This command gets all computer accounts.

### Example 6: Get all computers with a name starting with Computer01 or Computer02

```powershell
Get-ADComputer -Filter 'Name -like "Computer01*" -or Name -like "Computer02*"' -Properties IPv4Address |

>     Format-Table Name,DNSHostName,IPv4Address -AutoSize

Reviewer: ```suggestion
    Format-Table Name, DNSHostName, IPv4Address -AutoSize
```

```

```Output
name        dnshostname            ipv4address
----        -----------            -----------
Computer01-SRV1 Computer01-SRV1.Computer01.com 10.194.99.181
Computer02-SRV2 Computer02-SRV2.Computer02.com 10.194.100.3
```

### Example 7: Get all computers with a name starting with a string AND password last set before 30 days

```powershell
$Date = [DateTime]::Today.AddDays(-30)
Get-ADComputer -Filter 'Name -like "Computer01*" -and PasswordLastSet -ge $Date' -Properties IPv4Address | 

>     Format-Table Name,DNSHostName,IPv4Address -AutoSize

Reviewer: ```suggestion
    Format-Table Name, DNSHostName, IPv4Address -AutoSize
```

```

### -Credential

Specifies the user account credentials to use to perform this task. The default
credentials are the credentials of the currently logged on user unless the
cmdlet is run from an Active Directory module for Windows PowerShell provider
drive. If the cmdlet is run from such a provider drive, the account associated
with the drive is the default.

> To specify this parameter, you can type a user name, such as User1 or
> Domain01\User01 or you can specify a **PSCredential** object. If you specify a
> user name for this parameter, the cmdlet prompts for a password.

Reviewer: ```suggestion
To specify this parameter, you can type a user name, such as `User1` or `Domain01\User01` or you
can specify a **PSCredential** object. If you specify a user name for this parameter, the cmdlet
prompts for a password.
```

You can also create a **PSCredential** object by using a script or by using the

> **Get-Credential** cmdlet. You can then set the _Credential_ parameter to the

Reviewer: ```suggestion
`Get-Credential` cmdlet. You can then set the **Credential** parameter to the
```

**PSCredential** object.

> If the acting credentials do not have directory-level permission to perform the
> task, Active Directory module for Windows PowerShell returns a terminating
> error.

Reviewer: Opportunistic reword - it's the _cmdlet_ returning an error here, not the module.

```suggestion
If the acting credentials do not have directory-level permission to perform the task, the cmdlet
returns a terminating error.
```

```

### -Filter

Specifies a query string that retrieves Active Directory objects. This string
uses the Windows PowerShell Expression Language syntax. The Windows PowerShell

> Expression Language syntax provides rich type-conversion support for value types
> received by the _Filter_ parameter. The syntax uses an in-order representation,
> which means that the operator is placed between the operand and the value. For
> more information about the _Filter_ parameter, type `Get-Help`
> [about_ActiveDirectory_Filter](/previous-versions/windows/server/hh531527(v=ws.10)).

Reviewer: ```suggestion
Expression Language syntax provides rich type-conversion support for value types received by the
**Filter** parameter. The syntax uses an in-order representation, which means that the operator is
placed between the operand and the value. For more information about the **Filter** parameter, type
`Get-Help` [about_ActiveDirectory_Filter](/previous-versions/windows/server/hh531527(v=ws.10)).
```

For a list of supported types for \<value\>, type `Get-Help about_ActiveDirectory_ObjectModel`.

> Note: Windows PowerShell wildcards other than \*, such as ?, are not supported
> by the _Filter_ syntax.

Reviewer: We have a special syntax for notes - we should probably use it while we're updating.

```suggestion
> [!NOTE]
> Wildcards other than `*`, such as `?`, are not supported by the **Filter** syntax.
```

> Note: To query using LDAP query strings, use the _LDAPFilter_ parameter.

Reviewer: ```suggestion
> [!NOTE]
> To query using LDAP query strings, use the **LDAPFilter** parameter.
```

```

### -Identity

Specifies an Active Directory computer object by providing one of the following property values.
The identifier in parentheses is the LDAP display name for the attribute.
The acceptable values for this parameter are:

- A distinguished name

> - A GUID (objectGUID)
> - A security identifier (objectSid)
> - A Security Accounts Manager account name (sAMAccountName)

Reviewer: ```suggestion
- A GUID (`objectGUID`)
- A security identifier (`objectSid`)
- A Security Accounts Manager account name (`sAMAccountName`)
```

```

### -LDAPFilter

Specifies an LDAP query string that is used to filter Active Directory objects.

> You can use this parameter to run your existing LDAP queries. The _Filter_
> parameter syntax supports the same functionality as the LDAP syntax. For more
> information, see the _Filter_ parameter description or type `Get-Help`
> [about_ActiveDirectory_Filter](/previous-versions/windows/server/hh531527(v=ws.10)).

Reviewer: ```suggestion
You can use this parameter to run your existing LDAP queries. The **Filter**
parameter syntax supports the same functionality as the LDAP syntax. For more
information, see the **Filter** parameter description or type `Get-Help`
[about_ActiveDirectory_Filter](/previous-versions/windows/server/hh531527(v=ws.10)).
```

```

### -Partition

Specifies the distinguished name of an Active Directory partition.
The distinguished name must be one of the naming contexts on the current directory server.

> The cmdlet searches this partition to find the object defined by the _Identity_ parameter.

Reviewer: ```suggestion
The cmdlet searches this partition to find the object defined by the **Identity** parameter.
```

> In many cases, a default value is used for the _Partition_ parameter if no value

Reviewer: ```suggestion
In many cases, a default value is used for the **Partition** parameter if no value
```

is specified. The rules for determining the default value are given below. Note
that rules listed first are evaluated first and once a default value can be
determined, no further rules are evaluated.

In Active Directory Domain Services environments, a default value for

> _Partition_ is set in the following cases:

Reviewer: ```suggestion
**Partition** is set in the following cases:
```

> - If the _Identity_ parameter is set to a distinguished name, the default value
>   of _Partition_ is automatically generated from this distinguished name.
> - If running cmdlets from an Active Directory provider drive, the default value
>   of _Partition_ is automatically generated from the current path in the drive.
> - If none of the previous cases apply, the default value of _Partition_ is set
>   to the default partition or naming context of the target domain.

Reviewer: ```suggestion
- If the **Identity** parameter is set to a distinguished name, the default value of **Partition**
  is automatically generated from this distinguished name.
- If running cmdlets from an Active Directory provider drive, the default value of **Partition** is
  automatically generated from the current path in the drive.
- If none of the previous cases apply, the default value of **Partition** is set to the default
  partition or naming context of the target domain.
```

In Active Directory Lightweight Directory Services (AD LDS) environments, a

> default value for _Partition_ is set in the following cases:
> 
> - If the _Identity_ parameter is set to a distinguished name, the default value
>   of _Partition_ is automatically generated from this distinguished name.
> - If running cmdlets from an Active Directory provider drive, the default value
>   of _Partition_ is automatically generated from the current path in the drive.
> - If the target AD LDS instance has a default naming context, the default value
>   of _Partition_ is set to the default naming context. To specify a default
>   naming context for an AD LDS environment, set the
>   **msDS-defaultNamingContext** property of the Active Directory directory
>   service agent (DSA) object (**nTDSDSA**) for the AD LDS instance.
> - If none of the previous cases apply, the _Partition_ parameter will not take any default value.

Reviewer: ```suggestion
default value for **Partition** is set in the following cases:

- If the **Identity** parameter is set to a distinguished name, the default value of **Partition**
  is automatically generated from this distinguished name.
- If running cmdlets from an Active Directory provider drive, the default value of **Partition** is
  automatically generated from the current path in the drive.
- If the target AD LDS instance has a default naming context, the default value of **Partition** is
  set to the default naming context. To specify a default naming context for an AD LDS environment,
  set the **msDS-defaultNamingContext** property of the Active Directory directory service agent
  (DSA) object (**nTDSDSA**) for the AD LDS instance.
- If none of the previous cases apply, the **Partition** parameter will not take any default value.
```

```

### -Properties

Specifies the properties of the output object to retrieve from the server.
Use this parameter to retrieve properties that are not included in the default set.

Specify properties for this parameter as a comma-separated list of names.
To display all of the attributes that are set on the object, specify * (asterisk).

To specify an individual extended property, use the name of the property. For
properties that are not default or extended properties, you must specify the
LDAP display name of the attribute.

> To retrieve properties and display them for an object, you can use the Get-*
> cmdlet associated with the object and pass the output to the **Get-Member**
> cmdlet.

Reviewer: ```suggestion
To retrieve properties and display them for an object, you can use the `Get-*` cmdlet associated
with the object and pass the output to the `Get-Member` cmdlet.
```

```

### -SearchBase

Specifies an Active Directory path to search under.

When you run a cmdlet from an Active Directory provider drive, the default value
of this parameter is the current path of the drive.

When you run a cmdlet outside of an Active Directory provider drive against an
Active Directory Domain Services target, the default value of this parameter is
the default naming context of the target domain.

When you run a cmdlet outside of an Active Directory provider drive against an
AD LDS target, the default value is the default naming context of the target AD
LDS instance if one has been specified by setting the
**msDS-defaultNamingContext** property of the Active Directory directory service
agent object (**nTDSDSA**) for the AD LDS instance. If no default naming context
has been specified for the target AD LDS instance, then this parameter has no
default value.

> When the value of the _SearchBase_ parameter is set to an empty string and you
> are connected to a global catalog port, all partitions are searched. If the
> value of the _SearchBase_ parameter is set to an empty string and you are not
> connected to a global catalog port, an error is thrown.

Reviewer: ```suggestion
When the value of the **SearchBase** parameter is set to an empty string and you are connected to a
global catalog port, all partitions are searched. If the value of the **SearchBase** parameter is
set to an empty string and you are not connected to a global catalog port, an error is thrown.
```

```

### -Server

Specifies the Active Directory Domain Services instance to connect to, by
providing one of the following values for a corresponding domain name or
directory server. The service may be any of the following: Active Directory
Lightweight Domain Services, Active Directory Domain Services or Active
Directory snapshot instance.

Specify the Active Directory Domain Services instance in one of the following ways:

Domain name values:

- Fully qualified domain name
- NetBIOS name

Directory server values:

- Fully qualified directory server name
- NetBIOS name
- Fully qualified directory server name and port

The default value for this parameter is determined by one of the following
methods in the order that they are listed:

> - By using the _Server_ value from objects passed through the pipeline

Reviewer: ```suggestion
- By using the **Server** value from objects passed through the pipeline
```

```

### CommonParameters

This cmdlet supports the common parameters: -Debug, -ErrorAction,
-ErrorVariable, -InformationAction, -InformationVariable, -OutVariable,
-OutBuffer, -PipelineVariable, -Verbose, -WarningAction, and -WarningVariable.
For more information, see
[about_CommonParameters](https://go.microsoft.com/fwlink/?LinkID=113216).

## INPUTS

### None or Microsoft.ActiveDirectory.Management.ADComputer

> A computer object is received by the _Identity_ parameter.

Reviewer: ```suggestion
A computer object is received by the **Identity** parameter.
```

## OUTPUTS

### Microsoft.ActiveDirectory.Management.ADComputer

Returns one or more computer objects.

This Get-ADComputer cmdlet returns a default set of **ADComputer** property values.

> To retrieve additional **ADComputer** properties, use the _Properties_ parameter of this cmdlet.

Reviewer: ```suggestion
To retrieve additional **ADComputer** properties, use the **Properties** parameter of this cmdlet.
```

To view the properties for an **ADComputer** object, see the following examples.

> To run these examples, replace \<computer\> with a computer identifier such as

Reviewer: ```suggestion
To run these examples, replace `<computer>` with a computer identifier such as
```

`Get-ADComputer`\<computer\>`-Properties ALL | Get-Member`

## NOTES

> - This cmdlet does not work with AD LDS with its default schema. By default AD
>   LDS schema does not have a computer class, but if the schema is extended to
>   include it, this cmdlet will work with LDS.

Reviewer: ```suggestion
- This cmdlet doesn't work with AD LDS with its default schema. By default the AD LDS schema
  doesn't have a computer class, but if the schema is extended to include it, this cmdlet will work
  with LDS.
```
